File: Moderate/Backtracking/Subsets.java
/*
Input: nums = [1,2,3]
Output:
[
  [3],
  [1],
  [2],
  [1,2,3],
  [1,3],
  [2,3],
  [1,2],
  []
]

1 2 3 4**
1
2
1 2
3
1 2 3
4
1 2 3 4

2 1 3 4
2
*/
class Solution {
    public List<List<Integer>> subsets(int[] nums) {
		List<List<Integer>> result = new ArrayList<List<Integer>>();
		ArrayList<Integer> innerList = new ArrayList<Integer>();
		if(nums.length==0)
		{
			return result;
		}        
		subsets(nums,0,nums.length-1,result,innerList);
		return result;
    }
    
    private void subsets(int[] nums, int start, int end, List<List<Integer>> result, ArrayList<Integer> innerList)
    {
    	if(start<=end)
    	{
    		ArrayList tempList = (ArrayList<Integer>)innerList.clone();
			innerList.add(nums[start]);
			result.add(tempList);
			subsets(nums,start+1,end,result,innerList);
			innerList = tempList;
    	}
    }
}

File: CSharp/Medium/Backtracking/SubSets.cs
/*
Example:

Input: nums = [1,2,3]
Output:
[
  [3],
  [1],
  [2],
  [1,2,3],
  [1,3],
  [2,3],
  [1,2],
  []
]

1,2,3

{}
{3}
{2}
{2,3}
{1}
{1,2}
{1,2,3}
*/
public class Solution {
    public IList<IList<int>> Subsets(int[] nums) {
        IList<IList<int>> result = new List<IList<int>>();
        if(nums.Length==0)
        {
        	return result;
        }
        IList<int> innerList = new List<int>();
        Subset(nums,0,innerList,result);
        return result;
    }

    private void Subset(int[] nums, int index, IList<int> innerList, IList<IList<int>> result)
    {
    	if(index<nums.Length)
    	{
    		
    		IList<int> tempList = new List<int>();
    		for(int i=0;i<innerList.Count;i++)
    		{
    			tempList.Add(innerList[i]);
    		}
    		tempList.Add(nums[index]);
        Subset(nums,index+1,innerList,result);
    		Subset(nums,index+1,tempList,result);
    	}
    	else
    	{
    		result.Add(innerList);
    	}
    }
}